From b363a99b3e3e5ea510f62c7af382dd4db47590b9 Mon Sep 17 00:00:00 2001
From: Andreas Scherer <andreas_github@freenet.de>
Date: Wed, 10 Jul 2024 14:20:50 +0200
Subject: [PATCH 272/272] Resurrect the shared object idea.

Now that the 'extern' mess has been cleared, it's no longer a
problem to put the reusable MMIX modules into a shared object.

Even issue #9 doesn't rear its head on MacOS.
---
 Makefile | 27 +++++++++++++++++----------
 1 file changed, 17 insertions(+), 10 deletions(-)

diff --git a/Makefile b/Makefile
index 1a3c488..3c0edbb 100644
--- a/Makefile
+++ b/Makefile
@@ -6,7 +6,9 @@
 #   In fact, CWEB 3.61 is recommended for making hardcopy or PDF documentation.
 
 #   If you prefer optimization to debugging, change -g to something like -O:
-CFLAGS = -g
+CFLAGS = -g -fpic
+LDFLAGS = -L.
+LDLIBS = -lmmix
 
 #   Uncomment the second line if you use pdftex to bypass .dvi files:
 PDFTEX = dvipdfm
@@ -62,7 +64,7 @@ TESTFILES = *.mms silly.run silly.out *.mmconfig *.mmix
 MISCFILES = Makefile makefile.dos README mmix.mp mmix.1
 ALL = $(WEBFILES) $(TESTFILES) $(MISCFILES)
 
-basic:  mmixal mmix
+basic:  lib mmixal mmix
 
 doc:    mmix-doc.ps mmixal.dvi mmix-sim.dvi
 	dvips -pp 0-13 mmixal.dvi -o mmixal-intro.ps
@@ -73,6 +75,11 @@ all:    basic mmotype mmmix
 clean:
 	rm -f *~ *.o *.c *.h *.tex *.log *.dvi *.toc *.idx *.scn *.ps core
 
+lib: libmmix.so
+
+libmmix.so: mmix-arith.o mmix-config.o mmix-io.o mmix-pipe.o
+	$(CC) -shared $^ -o $@
+
 .SECONDEXPANSION:
 mmix-pipe.o mmix-sim.o: $$(subst .o,.c,$$@) mmix-io.c
 	perl -pe "s/(#define ABSTIME) \d*/\1 `date +%s`/" -i $<
@@ -80,17 +87,17 @@ mmix-pipe.o mmix-sim.o: $$(subst .o,.c,$$@) mmix-io.c
 
 mmix-config.c: mmix-pipe.c
 
-mmmix:  mmix-arith.o mmix-config.o mmix-io.o mmix-pipe.o mmmix.c
-	$(CC) $(CFLAGS) $^ -o $@
+mmmix:  mmmix.o lib
+	$(CC) $(LDFLAGS) $< -o $@ $(LDLIBS)
 
-mmixal: mmix-arith.o mmixal.c
-	$(CC) $(CFLAGS) $^ -o $@
+mmixal: mmixal.c lib
+	$(CC) $(LDFLAGS) $< -o $@ $(LDLIBS)
 
-mmix:   mmix-arith.o mmix-io.o mmix-sim.o
-	$(CC) $(CFLAGS) $^ -o $@
+mmix:   mmix-sim.o lib
+	$(CC) $(LDFLAGS) $< -o $@ $(LDLIBS)
 
-mmotype: mmotype.c mmix-arith.o
-	$(CC) $(CFLAGS) $^ -o $@
+mmotype: mmotype.c lib
+	$(CC) $(LDFLAGS) $< -o $@ $(LDLIBS)
 
 tarfile: $(ALL)
 	tar cvf /tmp/mmix.tar $(ALL)
-- 
2.45.2

